### PR TITLE
Use generator expressions in some built-in functions

### LINKED ISSUE #447: Use generator expressions in some built-in functions
From [PEP289](https://www.python.org/dev/peps/pep-0289/), in Python2.4+, some builtin functions could use generator expressions instead of list comprehensions, e.g.:

```diff
- sum([i for i in range(100)])
+ sum(i for i in range(100))
```
Likewise for `max` and `min`. Would this be a welcome addition?

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #447

### CODE REVIEW

File: pyupgrade/_plugins/generator_expressions_pep289.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    (
+        'all',
+        'any',
+        'bytearray',
+        'bytes',
+        'frozenset',
+        'list',
+        'max',
+        'min',
+        'print',
```
Comment (asottile) [inline pyupgrade/_plugins/generator_expressions_pep289.py:28]: print of a list is different than print of a generator

File: pyupgrade/_plugins/generator_expressions_pep289.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+ALLOWED_FUNCS = frozenset(
+    (
+        'all',
+        'any',
+        'bytearray',
+        'bytes',
+        'frozenset',
+        'list',
```
Comment (asottile) [inline pyupgrade/_plugins/generator_expressions_pep289.py:25]: set should be added too

File: pyupgrade/_plugins/generator_expressions_pep289.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from pyupgrade._ast_helpers import ast_to_offset
+from pyupgrade._data import register
+from pyupgrade._data import State
+from pyupgrade._data import TokenFunc
+from pyupgrade._token_helpers import find_closing_bracket
+from pyupgrade._token_helpers import find_comprehension_opening_bracket
+
+
+ALLOWED_FUNCS = frozenset(
+    (
```
Comment (asottile) [inline pyupgrade/_plugins/generator_expressions_pep289.py:19]: you don't have to do the ugly thing black does here, you can use `frozenset((` and then `))`

File: tests/features/generator_expressions_pep289_test.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+@pytest.mark.parametrize(
+    ('s',),
+    (
+        pytest.param(
+            'sum((i for i in range(3)))',
+            id='Parenthesised generator expression',
+        ),
+        pytest.param(
+            'sum([i for i in range(2)], x=2)',
```
Comment (asottile) [inline tests/features/generator_expressions_pep289_test.py:15]: this one would be fine to rewrite, but it'd have to keep the parens

**Comment by MarcoGorelli** [comment]
Thanks @asottile for your review, I've updated accordingly!

A note on `set` - this one's already covered by the other plugin which changes `set([i for i in range(3)])` into `{i for i in range(3)}`, so I don't think `set` should be included here as well - or should it?

File: pyupgrade/_plugins/generator_expressions_pep289.py (hunk 1 above)
Comment (MarcoGorelli) [reply pyupgrade/_plugins/generator_expressions_pep289.py]: Indeed, thanks - the "test" I wrote to make this list only checked the return types 😳 🤦

**Comment by asottile** [comment]
> 
> 
> Thanks @asottile for your review, I've updated accordingly!
> 
> A note on `set` - this one's already covered by the other plugin which changes `set([i for i in range(3)])` into `{i for i in range(3)}`, so I don't think `set` should be included here as well - or should it?

oh tru

**Review by asottile — APPROVED** [review]

**Pull request merged**